### PR TITLE
feat: RSA cryptosystem signature project

### LINKED ISSUE #101: Learn Cryptography By Building a Digital Signature System

### PR DESCRIPTION
**Pull request by vkWeb**
This is a simple implementation of RSA cryptosystem. The only caveat is that the implementation doesn't work with prime numbers greater than 10 like 11, 19. I have asked a question on [StackOverflow](https://crypto.stackexchange.com/questions/71334/rsa-digital-signature-system-is-failing-for-larger-prime-numbers) to deal with this.

@scissorsneedfoodtoo This is my first JavaScript project. So, I'll be looking forward to your feedback. I would love to see ways of improving the current implementation and project design. 😃

Closes #101

### CODE REVIEW

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo Actually, I realized just now, there's a small flaw in the implementation that might be the reason for prime > 10 issue. I'll fix it soon.

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo All issues fixed. Now, we are ready to break it down into units.

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo I would love to see your feedback on the test descriptions written so far. How can I make it better?

**Comment by scissorsneedfoodtoo** [comment]
@vkWeb, sure, let me take a look and make some suggestions.

File: rsa-cryptosystem-signature/index07.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  }
+  return hashValue;
+}
+
+function generatePrivateKey() {}
+
+function generatePublicKey() {}
+
+/*
+For different private keys, there's a different public key.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index07.js]: Each private key is paired with a public key.

Create a variable `publicKey` and set it to 0.

File: rsa-cryptosystem-signature/index08.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  }
+  return hashValue;
+}
+
+function generatePrivateKey() {}
+
+function generatePublicKey() {}
+
+/*
+Suppose Alice sends an encrypted message to Bob using a symmetric key. There's no way for Bob to verify if the sender is Alice. Since anyone with access to key can send Bob the message and claim to be Alice.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index08.js]: It would probably be better to build out one function at a time here, then we could work on breaking up this explanation.

Just so I understand how this process works, I'll try to summarize the process below. Please let me know if I'm missing anything!:
1. Alice wants to send a message to Bob, so she hashes and and encrypts a message using Bob's private key. She sends the unencrypted message, along with the hashed and encrypted message otherwise known as a signature.
2. Bob decrypts the signature which results in the original hashed message. He also hashes the unecrypted message he received and compares the hashed value to the decrypted signature.
3. If both match, then he's sure that the message is from Alice since she's the only one with his private key.

Is that how this process works?

Comment (vkWeb) [reply rsa-cryptosystem-signature/index08.js]: Yes, exactly that's how it works!

File: rsa-cryptosystem-signature/index01.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+/*
+Digital signature is a scheme to verify authenticity and maintain integrity of digital documents and messages.
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index01.js]: We should put disclaimer when we initialize primes.
 
I'll revise the rest of it.

File: rsa-cryptosystem-signature/index06.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+function hashTheMessage(message) {
+  let hashValue = 0;
+  for (let i = 0, msgLength = message.length; i < msgLength; ++i) {
+    hashValue += message.charCodeAt(i);
+  }
+  return hashValue;
+}
+
+/*
+In Caesar's cipher project, one key was used for both encryption and decryption. This is known as Symmetric key cryptography.
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index06.js]: It's important to introduce both keys together and thus the functions.
 
As per your advice, I'll revise the text.

File: rsa-cryptosystem-signature/index08.js (hunk 2 above)
Comment (vkWeb) [reply rsa-cryptosystem-signature/index08.js]: This explanation is the heart of the lesson. Campers must be able to grasp it properly. Should we improve the explanation? or it's fine?

Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index08.js]: @vkWeb, messed up the steps originally, so I just edited them above. :point_up:

Yes, I agree. It's important that we nail this explanation, but it's quite a lot to take in all at once. It would probably be best if we can break it up among several lessons somehow.

Comment (vkWeb) [reply rsa-cryptosystem-signature/index08.js]: @scissorsneedfoodtoo We should strictly explain it in a single lesson because as we progress, we'll focus more on RSA.

Still, I'll keep my eyes open ;)

Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index08.js]: Sure, just let me know. We're trying to keep lessons around 5 sentences or less, but there can be exceptions. The idea is that students will be working through these really fast, so everything should be as minimal as possible.

One idea I had is that you can explain parts of the process as students write the empty functions. Kind of a brief overview before you actually dive in and start coding. So you could have a lesson where students declare the empty `genereatePublicKey` function, and explain how each user should have a public key that can be used for encryption and decryption. Then the next lesson could be students declaring the empty `generatePrivateKey` function, and you explain that each user's public key is paired with a private key, and how they work together.

Perhaps you could continue the overview there, declaring empty functions around the editor and explaining how the whole process works before actually starting to code. What do you think?

Comment (vkWeb) [reply rsa-cryptosystem-signature/index08.js]: Yes, that empty function idea is my current approach of teaching. 

Umm. Okay, I'll break this explanation in parts and will see how things work.

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo I think you should look at test descriptions to simplify challenges a bit more. I'll appreciate your feedback on ways of improving the teaching approach.

And tell me from your perspective whether campers will be able to grasp the concepts through these tests or not. This project should be a hit one ;)

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo Do we need to break `index.html` and `style.css` into test descriptions? If yes then maybe we should turn this into a console based application to reduce unnecessary UI hassle.

**Comment by scissorsneedfoodtoo** [comment]
@vkWeb, yes, we're pushing for learners to write much if not all of the code for these projects. Turning this project into a console based application is a good idea and would help learners focus on learning about RSA signatures.

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo I have turned this into a console based application. The test descriptions are almost complete. I just need to revise some of them and add a bit more to illustrate hash collision.

Please provide your feedback so that I can further improve them.

File: rsa-cryptosystem-signature/index17.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+function hashTheMessage(message) {
+  let hashValue = 0;
+  for (let i = 0, msgLength = message.length; i < msgLength; ++i) {
+    hashValue += message.charCodeAt(i);
+  }
+  return hashValue;
+}
+
+/*
+If any integer till the `smallerNum` divide both the numbers then they are `not coprime`.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index17.js]: This might need to be broken down into two steps so we can better explain/review the remainder operator. Did my best to suggest some edits for now. Also, I'm sure my example code can be improved. Please feel free to change it in a future draft:

Divide `smallerNum` and `largerNum` by each of the numbers in the loop.
If `smallerNum` and `largerNum` can be divided evenly by any of those numbers, they are not co-prime.
Use an `if` statement to check if `smallerNum` and `largerNum` can be divided by `i` evenly.
Remember that the remainder (`%`)  operator can be used to check if one number is divisible by another:
```
const divisible = 9;
const notDivisible = 5;

console.log(divisible % 3); // 0
console.log(notDivisible % 2); // 1
```

File: rsa-cryptosystem-signature/index18.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+function hashTheMessage(message) {
+  let hashValue = 0;
+  for (let i = 0, msgLength = message.length; i < msgLength; ++i) {
+    hashValue += message.charCodeAt(i);
+  }
+  return hashValue;
+}
+
+/*
+Continuing with our example of 12 and 9. We observe, 3 divide both the numbers evenly. Thus, they are not coprime.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index18.js]: Going back to our example numbers 12 and 9, both can be divided by 3 evenly.
Thus, they are not co-prime.
Return `false` if `smallerNum` and `largerNum` can be divided by `i` evenly.

File: rsa-cryptosystem-signature/index46.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+function sendMsgToBob(message) {
+  const privateKey = generatePrivateKey();
+  generatePublicKey(privateKey);
+  const hashValue = hashTheMessage(message);
+  const generatedSignature = generateSignature(hashValue, privateKey);
+}
+
+/*
+Bob receives the data and signature. To verify the signature he first hashes the data through the same hash function which Alice used.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index46.js]: Bob receives the data and signature.

To verify the signature he first needs to hash the message with the same hash function Alice used.

Hash the received message and store the returned value as a variable the same way you did in the `sendMsgToBob()` function.

File: rsa-cryptosystem-signature/index47.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+function sendMsgToBob(message) {
+  const privateKey = generatePrivateKey();
+  generatePublicKey(privateKey);
+  const hashValue = hashTheMessage(message);
+  const generatedSignature = generateSignature(hashValue, privateKey);
+}
+
+/*
+If you remember, Alice hashed the message then he encrypted the hash value which we called signature.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index47.js]: Remember that Alice's signature is the hashed message which she then encrypted with her private key.

When we decrypt the signature using Alice's public key, we should get the original hash value of the data.

Call the `decryptSignature()` function and pass it `digitalSignature` as an argument.

Store the returned value as a variable named `decryptedSignature`.

File: rsa-cryptosystem-signature/index48.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+function sendMsgToBob(message) {
+  const privateKey = generatePrivateKey();
+  generatePublicKey(privateKey);
+  const hashValue = hashTheMessage(message);
+  const generatedSignature = generateSignature(hashValue, privateKey);
+}
+
+/*
+The decrypted signature is the hash value of the original data. If the hash value of recevied data and decrypted signature are equal then the signature is verified and data is intact.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index48.js]: The decrypted signature is the hash value of the original data.

If the hash value of received message and decrypted signature are equal then the signature is verified and the message is really from Alice.

Check for the equality of `hashValue` and `decryptedSignature`.

If they are equal then log a success message to the console.

File: rsa-cryptosystem-signature/index49.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+function sendMsgToBob(message) {
+  const privateKey = generatePrivateKey();
+  generatePublicKey(privateKey);
+  const hashValue = hashTheMessage(message);
+  const generatedSignature = generateSignature(hashValue, privateKey);
+}
+
+/*
+Log a failure message when the `hashValue` and `decryptedSignature` don't match since someone must have modified the data or maybe the signature was wrong.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index49.js]: Log a failure message when the `hashValue` and `decryptedSignature` don't match.

Either someone tampered with the message before Bob received it, or there was an error when Alice attached her signature.

File: rsa-cryptosystem-signature/index50.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+function generateSignature(hashValue, privateKey) {
+  return Math.pow(hashValue, privateKey) % N;
+}
+
+function decryptSignature(digitalSignature) {
+  return Math.pow(digitalSignature, publicKey) % N;
+}
+
+/*
+Call `sendAndVerify` function in `sendMsgToBob` and provide the `generatedSignature` and `message` as an argument.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index50.js]: Call the `sendAndVerify()` function and provide the `generatedSignature` and `message` as arguments.

File: rsa-cryptosystem-signature/index51.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  const decryptedSignature = decryptSignature(digitalSignature);
+  if (hashValue === decryptedSignature) {
+    console.log("Success! Data is intact and signature is verified.");
+  } else {
+    console.log("Failure! There's something wrong with data or signature.");
+  }
+}
+
+/*
+Call `sendMsgToBob` function and pass a super secret message as an argument. Observe the output of `console.log`.
```
Comment (scissorsneedfoodtoo) [inline rsa-cryptosystem-signature/index51.js]: Call the `sendMsgToBob()` function and pass it a super secret message as a string.

Observe the output of `console.log`.

File: rsa-cryptosystem-signature/index02.js [hunk 13]
```diff
@@ -0,0 +1,6 @@
+/*
+The fixed-size value is also known as a hash value.
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index02.js]: I thought of the same approach but it'll be a kinda spoon-feeding. This lesson will come after the JavaScript lesson so, we should let campers figure out where the code should go and I want them to make some mistakes ;)

File: rsa-cryptosystem-signature/index03.js [hunk 14]
```diff
@@ -0,0 +1,9 @@
+/* 
+Hash functions ideally use all of the data for hashing.
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index03.js]: Yes. Hash functions should use all of the data that's passed to it. 

I'll be editing some of the hashing descriptions. I'll include some more concepts.

File: rsa-cryptosystem-signature/index10.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+function generatePrivateKey() {}
+
+function generatePublicKey() {}
+
+function generateSignature() {}
+
+/* 
+Alice sends the signature along with the original data to Bob. Bob decrypts the received signature with Alice's public key and hashes the received data.
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index10.js]: @scissorsneedfoodtoo Amazing idea! Should I put there an Imgur link or upload the image file to the repo?

File: rsa-cryptosystem-signature/index17.js (hunk 5 above)
Comment (vkWeb) [reply rsa-cryptosystem-signature/index17.js]: @scissorsneedfoodtoo This lesson is at the end part of the curriculum so, I am expecting campers to understand the remainder operator.

File: rsa-cryptosystem-signature/index23.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
+const firstPrime = 2;
+const secondPrime = 5;
+const N = firstPrime * secondPrime;
+
+/*
+`Φ` function is multiplicative which means Φ(A * B) = Φ(A) * Φ(B).
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index23.js]: > Substituting the above values we get Φ(N) = (firstPrime - 1) * (secondPrime - 1)

The above statement may confuse some campers. We are suddenly using `N`.

We need to mention, `N = firstPrime * secondPrime`
Hence, `Φ(N) = (firstPrime - 1) * (secondPrime - 1)`

The original looks fine to me.

File: rsa-cryptosystem-signature/index24.js [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+  for (let i = 2; i <= smallerNum; ++i) {
+    if (smallerNum % i === 0 && largerNum % i === 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/*
+Private key has two constraints:
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index24.js]: Good job there ;)

File: rsa-cryptosystem-signature/index30.js [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+      return privateKey;
+    }
+  }
+
+  console.log("Private key can't be generated.");
+  return 0;
+}
+
+/*
+Public key must fulfil the condition: `(publicKey * privateKey) mod phiOfN = 1`
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index30.js]: > Note that the remainder (%) operator is also known as the modulo operator.

☝️ Not required.

File: rsa-cryptosystem-signature/index36.js [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+      return;
+    }
+    ++publicKey;
+  }
+
+  console.log("Public key can't be generated.");
+}
+
+/*
+To encrypt data using RSA we use a mathematical equation: `encryptedData = (Data ^ privateKey) % N`.
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index36.js]: See, I really don't know how or why this equation works. The inventors worked on this equation for around one year. It's related to Diffie hellman key exchange equation where this power thing reverses the operation of encryption or decryption whatever the case may be.

I'll read more about this equation and will see if I understand it's internal.

File: rsa-cryptosystem-signature/index38.js [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+  console.log("Public key can't be generated.");
+}
+
+function generateSignature(hashValue, privateKey) {
+  return Math.pow(hashValue, privateKey) % N;
+}
+
+/*
+To decrypt data using RSA we use a mathematical equation similar to that of encryption:
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index38.js]: @scissorsneedfoodtoo I am not at all satisfied with this part of the lesson :(
It's like we are telling campers to write the syntax. We are not able to explain the internals of the equation.

@scissorsneedfoodtoo Lets find a way to explain this equation in layman terms.

File: rsa-cryptosystem-signature/index02.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index02.js]: @vkWeb, yes, I know what you mean. We definitely want to avoid spoon feeding learners the answers! However the format will be to position the test descriptions near the line that needs to be changed. This should help us keep the text as minimal as possible. Don't worry, there are still many good ways to challenge learners as they work through this, like with the modulo operator mentioned below.

File: rsa-cryptosystem-signature/index03.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index03.js]: I see, that makes sense. Looking forward to seeing your new descriptions and concepts :+1:

File: rsa-cryptosystem-signature/index17.js (hunk 5 above)
Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index17.js]: @vkWeb, that's a good point. You're right that they should be familiar with this already, assuming they follow the curriculum. I was thinking more along the lines of having them check smallerNum in the first lesson, then write the check for biggerNum, but it should be fine to have them write both in the same step.

Can you think of another hint or example that doesn't give away so much of the answer? I'm all for providing the formula if it's complicated like some of the ones below, but I'm sure we can think of a way to guide the learners through this step.

File: rsa-cryptosystem-signature/index23.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index23.js]: Thanks for your input. I was trying to keep things as lean as possible, but it's not worth causing a bunch of confusion. Let's go with your original description here.

File: rsa-cryptosystem-signature/index24.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index24.js]: Thanks! Though I just interpreted what you wrote is all.

File: rsa-cryptosystem-signature/index36.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index36.js]: Okay, sounds good. I'll look into it, too.

File: rsa-cryptosystem-signature/index38.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (scissorsneedfoodtoo) [reply rsa-cryptosystem-signature/index38.js]: Sounds like a plan. It would be best if we can find a good way to explain how this works, even it's a very high level summary.

File: rsa-cryptosystem-signature/index02.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index02.js]: Okay. I'll shift descriptions to their appropriate places.

File: rsa-cryptosystem-signature/index38.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index38.js]: Yes. I'll read more about RSA.

File: rsa-cryptosystem-signature/index17.js (hunk 5 above)
Comment (vkWeb) [reply rsa-cryptosystem-signature/index17.js]: @scissorsneedfoodtoo Writing both the steps together will be the right way to go.

Use an if statement to check whether smallerNum and largerNum can be divided by i evenly. If i divides both the numbers then they aren't coprime.

☝️ I have just refined your description. How does it look?

File: rsa-cryptosystem-signature/index43.js [hunk 21]
```diff
@@ -0,0 +1,65 @@
+const firstPrime = 2;
+const secondPrime = 5;
+const N = firstPrime * secondPrime;
+const phiOfN = (firstPrime - 1) * (secondPrime - 1);
+let publicKey = 0;
+
+/*
+Hash value should be less than `N` for encryption and decryption to properly work.
```
[the remark this answers is not in the document]
Comment (vkWeb) [reply rsa-cryptosystem-signature/index43.js]: @scissorsneedfoodtoo I think the current flow is fine. Also, I'll try to explain why `hashValue` should be less than `N`.

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo Here's the diagram to illustrate the signing process:
RSA signing process

☝️ How does this look? Do you feel campers will comfortably understand the process through this diagram.

**Comment by scissorsneedfoodtoo** [comment]
@vkWeb, your diagram looks great! Makes it much easier to understand the whole process. The only suggestion I have is to alter the text slightly:

Encrypt hash value --> Encrypt the hash value using Alice's private key
Decrypt the signature --> Decrypt the signature using Alice's public key

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo I think now the descriptions look good. I have made some adjustments to the diagram on your advice.

RSA signing process diagram 

Also, I have added some descriptions to illustrate the hash collision. Now, we only need to explain the internals of the encryption and decryption equation.

@scissorsneedfoodtoo Yes, we can merge this if you think everything looks fine. Maybe we can explain the equations in a future PR.
 
@scissorsneedfoodtoo, @QuincyLarson what do you think about the current state of this project? Your feedback will help me in improving this project.

**Comment by scissorsneedfoodtoo** [comment]
@vkWeb, thank you for updating your diagram. It's much easier to reason about now that the keys are in there. I just noticed a couple of things:

* Encrypt hash value using Alice's private key --> Encrypt the hash value using Alice's private key
* If the hash value and decrypted signature match then the signature is authentic and the message must be from Alice only --> If the hash value and decrypted signature match, then the signature is authentic and the message can only be from Alice / If the hash value and decrypted signature match, then the signature is authentic and the message must be from Alice

I think your project is definitely headed in the right direction. A few of the later lessons still need some editing, so I'll leave comments here and they can be handled in a future PR. Other than that, I'll keep trying to learn more about how those equations so hopefully we can explain them simply and thoroughly.

**Comment by vkWeb** [comment]
@scissorsneedfoodtoo Thanks for merging brother! 🎉 

Yes, we'll keep on improving the project in future PR(s). Btw please re-open the issue #101 and move it to `Coding Project Tests` in the kanban board.
